server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java

```diff
@@ -1023,7 +1038,11 @@ public void onResponse(ClusterState clusterState) {
                         public void onFailure(Exception e) {
                             assertEquals(testContextHeaderValue, threadContext.getHeader(testContextHeaderName));
                             assertEquals(List.of(testResponseHeaderValue), threadContext.getResponseHeaders().get(testResponseHeaderName));
-                            assertThat(e, instanceOf(FailedToCommitClusterStateException.class));
+                            if (expectFailure) {
+                                assertThat(e, instanceOf(ElasticsearchException.class));
```

Comment:
This is debatable, but preserves today's behaviour. Maybe we should return a `FailedToCommitClusterStateException` if the commit failed?

Reply:
Agreed, I think that is the right thing to do, since otherwise, a conflict could still be reported without the offender being committed. Also, I sort of think the failed to commit exception trumps the "application" exception here.

Reply:
Ok, I adjusted this in 4fc44103ddb867100e7b4d05ef17186369c1c50a.